src/v/cloud_storage/remote_partition.cc

```diff
@@ -339,67 +394,45 @@ class partition_record_batch_reader_impl final
         if (_reader->is_eof()) {
-            // move to the next segment
-            _partition->maybe_sync_with_manifest();
-            vlog(_ctxlog.debug, "maybe_reset_stream condition triggered");
-            _it++;
-            // We're comparing to the cached _end instead of
-            // _partition->_segments.end() to avoid the following pitfall. If
-            // the currently referenced segment has no data batches and we get
-            // to this point, but at the same time the new segment with the same
-            // base_offset (in kafka terms) was added to the map this new
-            // segment will replace the current one and we won't read it (since
-            // we're already done with the segment). This isn't a problem since
-            // we will just stop and the next fetch request will be able to read
-            // the data. But if not one but two or more segments were added to
-            // the _segments map we will just skip the segment with the same
-            // base_offset and proceed with the next one. The client will see a
-            // gap. The caching of the _segments.end() prevents this.
-            if (_it == _end) {
-                // This branch might be taken if the last segment was merged
-                // with the previous one. (TODO: fix this)
-                co_await set_end_of_stream();
-            } else {
-                // reuse config but replace the reader
-                auto config = _reader->config();
-                _partition->materialized().evict_reader(std::move(_reader));
-                vlog(_ctxlog.debug, "initializing new segment reader");
-                // It's safe to dereference '_it' since we just incremented it.
-                // If the underlying segment was deleted the _it will be equal
-                // to _end. But just in case if anyone will insert an
-                // asynchronous operation in between the increment and
-                // dereferencing we're checking the state of the iterator.
+            auto prev_max_offset = _reader->max_rp_offset();
+            auto config = _reader->config();
+            vlog(
+              _ctxlog.debug,
+              "maybe_reset_stream condition triggered after offset: {}, "
+              "reader's current log offset: {}, config.start_offset: {}, "
+              "reader's max log offset: {}, is EOF: {}, next base_offset "
+              "estimate: {}",
+              prev_max_offset,
+              _reader->current_rp_offset(),
+              config.start_offset,
+              _reader->max_rp_offset(),
+              _reader->is_eof(),
+              _next_offset);
+            _partition->materialized().evict_reader(std::move(_reader));
+            vlog(
+              _ctxlog.debug,
+              "initializing new segment reader {}, next offset",
+              config.start_offset,
+              _next_offset);
+            if (_next_offset != model::offset{}) {
```

Comment:
What if we were reading from the last segment when we started (so `_next_offset` is empty) but the next segment appeared while we were reading? Is it okay to stop reading in this case?